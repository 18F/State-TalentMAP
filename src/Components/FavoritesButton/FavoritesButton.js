import React, { Component } from 'react';
import PropTypes from 'prop-types';

class FavoritesButton extends Component {
  constructor(props) {
    super(props);
    this.state = {
      saved: false,
      localStorageKey: null,
    };
  }

  componentWillMount() {
    const localStorageKey = `${this.props.refKey}_${this.props.type}`;
    this.setState({ localStorageKey });
  }

  componentDidMount() {
    const isSaved = localStorage.getItem( // eslint-disable-line no-undef
      this.state.localStorageKey);
    if (isSaved) {
      const saved = true;
      this.initSetSaved(saved);
    }
  }

  getSavedState() {
    return this.state.saved;
  }

  initSetSaved(val) {
    const saved = val;
    this.setState({ saved });
  }

  toggleSaved() {
    if (this.state.saved) {
      localStorage.removeItem(this.state.localStorageKey); // eslint-disable-line no-undef
    } else {
      localStorage.setItem(this.state.localStorageKey, true); // eslint-disable-line no-undef
    }
    this.setState({ saved: !this.state.saved });
  }

  render() {
    let fromText = null;
    fromText = this.props.type === 'fav' ? 'Favorites' : fromText;
    fromText = this.props.type === 'compare' ? 'Comparison' : fromText;
    const buttonClass = this.state.saved ? 'usa-button-secondary' : '';
    const buttonText = this.getSavedState() ? `Remove from ${fromText}` : `Add to ${fromText}`;
    return (
      <div>
        <button id="changeSaved" className={buttonClass} onClick={() => this.toggleSaved()}>
          {buttonText}
        </button>
      </div>
    );
  }
}

FavoritesButton.propTypes = {
  refKey: PropTypes.string.isRequired,
  type: PropTypes.string.isRequired,
};

export default FavoritesButton;
