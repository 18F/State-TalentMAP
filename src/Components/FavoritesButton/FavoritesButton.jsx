import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { localStorageFetchValue, localStorageToggleValue } from '../../utilities';
import { EMPTY_FUNCTION } from '../../Constants/PropTypes';

class FavoritesButton extends Component {
  constructor(props) {
    super(props);
    this.state = {
      saved: false,
      localStorageKey: null,
      count: 0,
    };
  }

  componentWillMount() {
    const localStorageKey = this.props.type;
    this.setState({ localStorageKey });
  }

  componentDidMount() {
    this.getSaved();
  }

  onToggle() {
    this.props.onToggle();
  }

  getSaved() {
    const saved = localStorageFetchValue(this.state.localStorageKey, this.props.refKey);
    this.setState({ saved: saved.exists, count: saved.count });
  }

  getSavedState() {
    return this.state.saved;
  }

  exceedsLimit() {
    return this.state.count >= this.props.limit;
  }

  toggleSaved() {
    localStorageToggleValue(this.state.localStorageKey, this.props.refKey);
    this.setState({ saved: !this.state.saved });
    this.onToggle();
  }

  render() {
    let fromText = null;
    if (this.props.type === 'fav') {
      fromText = 'Favorites';
    } else if (this.props.type === 'compare') {
      fromText = 'Comparison';
    }
    let buttonClass = this.state.saved ? 'usa-button-secondary' : '';
    buttonClass = this.exceedsLimit() && !this.state.saved ? 'usa-button-disabled' : buttonClass;
    const buttonText = this.getSavedState() ? `Remove from ${fromText}` : `Add to ${fromText}`;
    const disabled = this.exceedsLimit() && !this.state.saved;
    return (
      <div>
        <button
          disabled={disabled}
          className={buttonClass}
          onClick={() => this.toggleSaved()}
        >
          {buttonText}
        </button>
      </div>
    );
  }
}

FavoritesButton.propTypes = {
  refKey: PropTypes.string.isRequired,
  type: PropTypes.string.isRequired,
  limit: PropTypes.number,
  onToggle: PropTypes.func,
};

FavoritesButton.defaultProps = {
  limit: 99,
  onToggle: EMPTY_FUNCTION,
};

export default FavoritesButton;
