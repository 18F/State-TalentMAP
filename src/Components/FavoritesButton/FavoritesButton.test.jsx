import { shallow } from 'enzyme';
import React from 'react';
import TestUtils from 'react-dom/test-utils';
import FavoritesButton from './FavoritesButton';

describe('FavoritesButton', () => {
  let favoritesButton = null;

  beforeEach(() => {
    favoritesButton = TestUtils.renderIntoDocument(<FavoritesButton refKey="0036" type="fav" />);
  });

  it('is defined', () => {
    expect(favoritesButton).toBeDefined();
  });

  it('it can accept different kinds of props', () => {
    const favoritesButtonCompare = shallow(
      <FavoritesButton refKey="0037" type="compare" />,
     );
    expect(favoritesButtonCompare).toBeDefined();
    const favoritesButtonOther = shallow(
      <FavoritesButton refKey="0037" type="other" />,
     );
    expect(favoritesButtonOther).toBeDefined();
  });

  it('can add a favorite', () => {
    const wrapper = shallow(<FavoritesButton refKey="0036" type="fav" />);
    wrapper.find('button').simulate('click');
    expect(wrapper.instance().state.saved).toBe(true);
  });

  it('can add and remove a favorite', () => {
    const wrapper = shallow(<FavoritesButton refKey="0037" type="fav" />);
    wrapper.find('button').simulate('click');
    expect(wrapper.instance().state.saved).toBe(true);
    wrapper.find('button').simulate('click');
    expect(wrapper.instance().state.saved).toBe(false);
  });

  it('can handle len key in state', () => {
    const wrapper = shallow(<FavoritesButton refKey="0037" type="fav" />);
    wrapper.instance().state.len = 100000; // greater than default limit
    wrapper.find('button').simulate('click');
    expect(wrapper.instance().state.saved).toBe(true);
  });
});
