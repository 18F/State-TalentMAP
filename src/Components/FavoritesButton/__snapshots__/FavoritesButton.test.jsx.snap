// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FavoritesButton can handle a disabled state 1`] = `
<button
  className="usa-button-disabled"
  disabled={true}
  onClick={[Function]}
>
  Add to Favorites
</button>
`;
